quickwit-search/src/query_runner.rs

```diff
@@ -0,0 +1,184 @@
+                search_error,
+                split_ids: vec![],
+                node_addr: client.grpc_addr(),
+            })
+    }
+
+    async fn query_plan(&self, request: Self::RootRequest) -> anyhow::Result<ExecutionPlan<Self::LeafRequest>> {
+        let split_metadata_list = list_relevant_splits(&request, self.metastore.as_ref()).await?;
+        let split_metadata_map: HashMap<String, SplitMetadata> = split_metadata_list
+            .into_iter()
+            .map(|split_metadata| (split_metadata.split_id.clone(), split_metadata))
+            .collect();
+        let leaf_search_jobs: Vec<Job> =
+            job_for_splits(&split_metadata_map.keys().collect(), &split_metadata_map);
+        let assigned_leaf_search_jobs = self.client_pool
+            .assign_jobs(leaf_search_jobs, &HashSet::default())
+            .await?;
+        let requests = assigned_leaf_search_jobs.iter()
+            .map(|(client, jobs)| {
+                let mut request_with_offset_0 = request.clone();
+                request_with_offset_0.start_offset = 0;
+                request_with_offset_0.max_hits += request.start_offset;
+                let leaf_request = LeafSearchRequest {
+                    search_request: Some(request_with_offset_0),
+                    split_ids: jobs.iter().map(|job| job.split.clone()).collect(),
+                };
+                (client.clone(), leaf_request)
+            })
+            .collect_vec();
+        Ok(ExecutionPlan::<Self::LeafRequest>{ requests })
+    }
+}
+
+struct FetchDocsExecutor {
+    client_pool: Arc<SearchClientPool>
+}
+
+#[async_trait]
+impl DistributedQueryRunner for FetchDocsExecutor {
+    type RootRequest = FetchDocsRequest;
+    type LeafRequest = FetchDocsRequest;
+    type LeafResult = FetchDocsResult;
+
+    async fn execute_leaf(&self, mut client: SearchServiceClient, leaf_request: Self::LeafRequest) -> Result<Self::LeafResult, NodeAddrSearchError> {
+        client.fetch_docs(leaf_request)
+            .await
+            .map_err(|search_error| NodeAddrSearchError {
+                search_error,
+                split_ids: vec![],
+                node_addr: client.grpc_addr(),
+            })
+    }
+
+    async fn query_plan(&self, request: Self::RootRequest) -> anyhow::Result<ExecutionPlan<Self::LeafRequest>> {
+        let mut partial_hits_map: HashMap<String, Vec<PartialHit>> = HashMap::new();
+        for partial_hit in request.partial_hits.iter() {
+            partial_hits_map
+                .entry(partial_hit.split_id.clone())
+                .or_insert_with(Vec::new)
+                .push(partial_hit.clone());
+        }
+        let fetch_docs_req_jobs = request.partial_hits
+            .iter()
+            .map(|hit| Job {
+                split: hit.split_id.to_string(),
+                cost: 1,
+            })
+            .collect_vec();
+        let doc_fetch_jobs = self.client_pool
+            .assign_jobs(fetch_docs_req_jobs, &HashSet::default())
```

Comment:
The logic we have today is not like that, is it? Today we reassign the job to the same leaf that returned the result I think